### DISCUSSION TITLE
System color theme not working with next js?

### DISCUSSION THREAD

**Discussion by nekmart** [mantinedev/mantine#2772] [Q&A]
### What package has an issue

@mantine/core

### Describe the bug

Hello, I want to use mantine for my next `nextjs` project. That's why I am doing practice with mantine. Here I am getting one issue with mantine dark theme. Here in documentation I find this with next js cookie based color theme- https://mantine.dev/guides/dark-theme/#save-color-scheme-in-cookie.

But this example do not provide about system color theme integration in nextjs. I have found another docs- https://mantine.dev/guides/dark-theme/#detect-user-preferred-color-scheme. After reading this docs I am try to apply `preferredColorScheme` in this way. 

```
// _app.tsx file
import { GetServerSidePropsContext } from 'next';
import { useState } from 'react';
import { AppProps } from 'next/app';
// install cookies-next package to manage cookies
import { getCookie, setCookie } from 'cookies-next';
import { MantineProvider, ColorScheme, ColorSchemeProvider } from '@mantine/core';
import { useColorScheme } from '@mantine/hooks';

export default function App(props: AppProps & { colorScheme: ColorScheme }) {
  const { Component, pageProps } = props;
  const preferredColorScheme = useColorScheme();
  const [colorScheme, setColorScheme] = useState<ColorScheme>(props.colorScheme || preferredColorScheme );

  const toggleColorScheme = (value?: ColorScheme) => {
    const nextColorScheme = value || (colorScheme === 'dark' ? 'light' : 'dark');
    setColorScheme(nextColorScheme);
    // when color scheme is updated save it to cookie
    setCookie('mantine-color-scheme', nextColorScheme, { maxAge: 60 * 60 * 24 * 30 });
  };

  return (
    <ColorSchemeProvider colorScheme={colorScheme} toggleColorScheme={toggleColorScheme}>
      <MantineProvider theme={{ colorScheme }} withGlobalStyles withNormalizeCSS>
        <Component {...pageProps} />
      </MantineProvider>
    </ColorSchemeProvider>
  );
}
App.getInitialProps = ({ ctx }: { ctx: GetServerSidePropsContext }) => ({
  // get color scheme from cookie
  colorScheme: getCookie('mantine-color-scheme', ctx),
});
```

But it is not working as well. I am also getting a flashed flickering, when dark mode is selected. Can I get a flashed flicker free dark mode experience.

### What version of @mantine/hooks page do you have in package.json?

latest version (5.6.0)

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

No

### Possible fix

As we don't know browser theme in server side, here it can be done with data attribute in html root element. Actually I am not a pro developer, but I think it can be done as we want no-flashing dark mode experience. That's why I can refer some root package like - [next-dark-mood](https://github.com/xeoneux/next-dark-mode), [next-theme-mode](https://www.npmjs.com/package/next-theme-mode), [next-themes](https://www.npmjs.com/package/next-themes)

You can see they all provide a no-flashing dark mode even when it selected system mode.

**Comment by rtivital** [comment]
There is no way to prevent flickering without using cookie – it is not possible to get preferred color scheme on server. If you want to learn more, you can browse through similar discussions, for example https://github.com/mantinedev/mantine/discussions/1584

**Reply by nekmart** [reply]
Can I prevent flickering by using cookie? And here Can I use preferred color scheme?

**Reply by nekmart** [reply]
Actually, I want to use two things together. I want to use cookies based dark theme with preferred color scheme.

**Reply by rtivital** [reply]
You can use cookies with preferred color scheme, when application mounts save color scheme to cookie, then on each next request color scheme will be the same as user's preference.

**Reply by siamahnaf** [reply]
Can you give me a example? Actually I am also facing this issue.

**Reply by rtivital** [reply]
Do not have an example, but the logic should be something like this:

```tsx
const [colorScheme, setColorScheme] = useState(initialColorScheme || 'light');
useEffect(() => {
  const preferredColorScheme = /* get user preferred color scheme */;
  setColorScheme(preferredColorScheme);
  setCookie('color-scheme', preferredColorScheme);
}, []);
```

This way:
- You will receive user preferred color scheme in initial value on second render
- There won't be any flickers on second and next requests unless the user changes preferred color scheme
